pkg/virt-controller/watch/migration.go

```diff
@@ -49,84 +70,183 @@ func (md *MigrationDispatch) Execute(store cache.Store, queue workqueue.RateLimi
-				logger.Error().Msgf("VM with name %s is in state %s, no migration possible. Marking migration as failed", vm.GetObjectMeta().GetName(), vm.Status.Phase)
-				setMigrationFailed(migrationCopy)
-				return
 			}
+		}
+
+		// Let's check if the job already exists, it can already exist in case we could not update the VM object in a previous run
+		migrationPod, exists, err := md.vmService.GetMigrationJob(migration)
+
+		if err != nil {
+			logger.Error().Reason(err).Msg("Checking for an existing migration job failed.")
+			queue.AddRateLimited(key)
+			return
+		}
 
-			if err := mergeConstraints(migration, vm); err != nil {
-				logger.Error().Reason(err).Msg("merging Migration and VM placement constraints failed.")
+		if !exists {
+			sourceNode, err := md.clientset.CoreV1().Nodes().Get(vm.Status.NodeName, metav1.GetOptions{})
+			if err != nil {
+				logger.Error().Reason(err).Msgf("Fetching source node %s failed.", vm.Status.NodeName)
 				queue.AddRateLimited(key)
 				return
 			}
-			podList, err := md.vmService.GetRunningVMPods(vm)
+			targetNode, err := md.clientset.CoreV1().Nodes().Get(vm.Status.MigrationNodeName, metav1.GetOptions{})
 			if err != nil {
-				logger.Error().Reason(err).Msg("could not fetch a list of running VM target pods")
+				logger.Error().Reason(err).Msgf("Fetching target node %s failed.", vm.Status.MigrationNodeName)
 				queue.AddRateLimited(key)
 				return
 			}
 
-			numOfPods, migrationPodExists := investigateTargetPodSituation(migration, podList)
-			if numOfPods > 1 && !migrationPodExists {
-				logger.Error().Reason(err).Msg("another migration seems to be in progress, marking Migration as failed")
-				// Another migration is currently going on
-				setMigrationFailed(migrationCopy)
+			if err := md.vmService.StartMigration(migration, vm, sourceNode, targetNode, targetPod); err != nil {
+				logger.Error().Reason(err).Msg("Starting the migration job failed.")
+				queue.AddRateLimited(key)
 				return
-			} else if numOfPods == 1 && !migrationPodExists {
-				// We need to start a migration target pod
-				// TODO, this detection is not optimal, it can lead to strange situations
-				err := md.vmService.SetupMigration(migration, vm)
-				if err != nil {
-					logger.Error().Reason(err).Msg("creating a migration target node failed")
-					queue.AddRateLimited(key)
-					return
-				}
 			}
-			logger.Error().Msg("another migration seems to be in progress, marking Migration as failed")
-			migrationCopy.Status.Phase = v1.MigrationInProgress
-			// TODO indicate when this has happened
-			err = md.vmService.UpdateMigration(migrationCopy)
-			if err != nil {
-				logger.Error().Reason(err).Msgf("updating migration state failed : %v ", err)
+			//TODO add a new state that more accurately reflects the process up the this point
+			//and then use MigrationInProgress to accurately indicate the actual migration pod is running
+			//setMigrationPhase(migration, v1.MigrationInProgress)
```

Comment:
MigrationTargetRequested when we call to start the target pod.  MigrationRunning will be when we call to start the migration process itself.